README.md
# go-services-qingstor

[![Build Status](https://github.com/beyondstorage/go-service-qingstor/workflows/Unit%20Test/badge.svg?branch=master)](https://github.com/beyondstorage/go-service-qingstor/actions?query=workflow%3A%22Unit+Test%22)
[![License](https://img.shields.io/badge/license-apache%20v2-blue.svg)](https://github.com/Xuanwo/storage/blob/master/LICENSE)
[![go-service-qingstor](https://img.shields.io/matrix/go-service-qingstor:aos.dev.svg?server_fqdn=chat.aos.dev&label=%23go-service-qingstor%3Aaos.dev&logo=matrix)](https://matrix.to/#/#go-service-qingstor:aos.dev)

[QingStor Object Storage](https://www.qingcloud.com/products/objectstorage/) services support for [go-storage](https://github.com/beyondstorage/go-storage)

## Install

```go
go get github.com/beyondstorage/go-service-qingstor/v3
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/)
and this project adheres to [Semantic Versioning](https://semver.org/).

## [v3.2.0] - 2021-07-22

### Added

- ci: Add gofmt action (#62)
- ci: Add diff check action (#65)
- ci: Add dependabot auto build support (#66)

### Changed

- storage: Update types in service.toml to golang types (#71)
- storage: Update append as described in GSP-134 (#71)
- storage: Update list as described in GSP-654 (#71)
- build(deps): Migrate to go-endpoint (#74)

### Fixed

- ci: Fix auto-build not work correctly
- storage: Fix invalid argument for copy and move (#72)
- storage: Fix append behavior (#73)

### Upgraded

- build(deps): Bump github.com/google/uuid from 1.2.0 to 1.3.0 (#61)

## [v3.1.0] - 2021-06-29

### Added

- *: Implement GSP-87 Feature Gates (#53)
- storage: Implement GSP-93 Add ObjectMode Pair (#58)
- storage: Implement GSP-97 Add Restrictions In Storage Metadata (#58)

### Changed

- *: Implement GSP-109: Redesign Features (#58)
- *: Implement GSP-117 Rename Service to System as the Opposite to Global (#58)

### Upgraded

- build(deps): bump github.com/golang/mock from 1.5.0 to 1.6.0 (#56)

## [v3.0.0] - 2021-05-24

### Added

- storage: Add appender support (#40)
- *: Implement GSP-47 & GSP-51 (#46)
- storage: Implement GSP-61 Add object mode check for operations (#49)

### Changed

- service: Use path style instead of vhost (#43)
- service: Fix location not detected correctly (#45)
- storage: Idempotent storager delete operation (#44)
- storage: Implement GSP-62 WriteMultipart returns Part (#47)
- storage: Check if part number is valid when multipart upload (#48)
- *: Implement GSP-73 Organization rename (#51)

## [v2.1.0] - 2021-04-24

### Added

- *: Implement proposal unify object metadata (#25)
- storage: Normalize iterator next function names (#27)
- pair: Implement default pair support for service (#29)
- *: Set default pair when init (#31)
- storage: Implement Create API (#33)
- storage: Set multipart attributes when create multipart (#34)
- *: Add UnimplementedStub (#35)
- storage: Implement SSE support (#37)
- tests: Introduce STORAGE_QINGSTOR_INTEGRATION_TEST (#39)
- storage: Implement GSP-40 (#41)

### Changed

- storage: Clean up next page logic
- build: Make sure integration tests has been executed
- docs: Migrate zulip to matrix
- docs: Remove zulip
- ci: Only run Integration Test while push to master
- storage: Rename SSE related pairs to meet GSP-38 (#38)

### Fixed

- storage: Fix multipart integration tests (#36)

### Removed

- *: Remove parsed pairs pointer (#28)

### Upgrade

- build(deps): bump github.com/qingstor/qingstor-sdk-go/v4 (#26)

## [v2.0.0] - 2021-01-17

### Added

- tests: Add integration tests (#17)
- storage: Implement Fetcher (#19)
- storage: Implement proposal Unify List Operation (#20)
- *: Implement Segment API Redesign (#21)
- storage: Implement proposal Object Mode (#22)

### Changed

- Migrate to go-storage v3 (#23)

## v1.0.0 - 2020-11-12

### Added

- Implement qingstor services.

[v3.2.0]: https://github.com/beyondstorage/go-service-qingstor/compare/v3.1.0...v3.2.0
[v3.1.0]: https://github.com/beyondstorage/go-service-qingstor/compare/v3.0.0...v3.1.0
[v3.0.0]: https://github.com/beyondstorage/go-service-qingstor/compare/v2.1.0...v3.0.0
[v2.1.0]: https://github.com/beyondstorage/go-service-qingstor/compare/v2.0.0...v2.1.0
[v2.0.0]: https://github.com/beyondstorage/go-service-qingstor/compare/v1.0.0...v2.0.0


utils_test.go
package qingstor

import (
	"errors"
	"testing"

	"github.com/golang/mock/gomock"
	"github.com/google/uuid"
	"github.com/pengsrc/go-shared/convert"
	qserror "github.com/qingstor/qingstor-sdk-go/v4/request/errors"
	"github.com/qingstor/qingstor-sdk-go/v4/service"
	"github.com/stretchr/testify/assert"

	"github.com/beyondstorage/go-endpoint"
	"github.com/beyondstorage/go-storage/v4/pairs"
	"github.com/beyondstorage/go-storage/v4/pkg/credential"
	"github.com/beyondstorage/go-storage/v4/services"
)

func Test_New(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	// Missing required pair
	_, _, err := New()
	assert.Error(t, err)
	assert.True(t, errors.Is(err, services.ErrRestrictionDissatisfied))

	// Valid case
	accessKey := uuid.New().String()
	secretKey := uuid.New().String()
	host := uuid.New().String()
	name := uuid.New().String()
	port := 1234
	srv, store, err := New(
		pairs.WithCredential(credential.NewHmac(accessKey, secretKey).String()),
		pairs.WithEndpoint(endpoint.NewHTTP(host, port).String()),
		pairs.WithLocation("test"),
		pairs.WithName(name),
	)
	assert.NoError(t, err)
	assert.NotNil(t, srv)
	assert.NotNil(t, store)
}

func TestIsBucketNameValid(t *testing.T) {
	tests := []struct {
		name string
		args string
		want bool
	}{
		{"start with letter", "a-bucket-test", true},
		{"start with digit", "0-bucket-test", true},
		{"start with strike", "-bucket-test", false},
		{"end with strike", "bucket-test-", false},
		{"too short", "abcd", false},
		{"too long (64)", "abcdefghijklmnopqrstuvwxyz123456abcdefghijklmnopqrstuvwxyz123456", false},
		{"contains illegal char", "abcdefg_1234", false},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := IsBucketNameValid(tt.args); got != tt.want {
				t.Errorf("IsBucketNameValid() = %v, want %v", got, tt.want)
			}
		})
	}
}

func TestGetAbsPath(t *testing.T) {
	cases := []struct {
		name         string
		base         string
		path         string
		expectedPath string
	}{
		{"under root", "/", "abc", "abc"},
		{"under prefix", "/root", "/abc", "root/abc"},
		{"under prefix ending with /", "/root/", "abc", "root/abc"},
		{"under unexpected prefix", "//abc", "/def", "/abc/def"},
	}

	for _, tt := range cases {
		t.Run(tt.name, func(t *testing.T) {
			client := Storage{workDir: tt.base}

			gotPath := client.getAbsPath(tt.path)
			assert.Equal(t, tt.expectedPath, gotPath)
		})
	}
}

func TestGetRelPath(t *testing.T) {
	cases := []struct {
		name         string
		base         string
		path         string
		expectedPath string
	}{
		{"under root", "/", "abc", "abc"},
		{"under prefix", "/root", "root/abc", "/abc"},
		{"under prefix ending with /", "/root/", "root/abc", "abc"},
		{"under unexpected prefix", "//abc", "/abc/def", "/def"},
	}

	for _, tt := range cases {
		t.Run(tt.name, func(t *testing.T) {
			client := &Storage{workDir: tt.base}

			gotPath := client.getRelPath(tt.path)
			assert.Equal(t, tt.expectedPath, gotPath)
		})
	}
}

func TestHandleQingStorError(t *testing.T) {
	{
		tests := []struct {
			name     string
			input    *qserror.QingStorError
			expected error
		}{
			{
				"not found",
				&qserror.QingStorError{
					StatusCode:   404,
					Code:         "",
					Message:      "",
					RequestID:    "",
					ReferenceURL: "",
				},
				services.ErrObjectNotExist,
			},
		}

		for _, tt := range tests {
			t.Run(tt.name, func(t *testing.T) {
				assert.True(t, errors.Is(formatError(tt.input), tt.expected))
			})
		}
	}

	{
		tests := []struct {
			name     string
			input    *qserror.QingStorError
			expected error
		}{
			{
				"permission_denied",
				&qserror.QingStorError{
					StatusCode:   403,
					Code:         "permission_denied",
					Message:      "",
					RequestID:    "",
					ReferenceURL: "",
				},
				services.ErrPermissionDenied,
			},
			{
				"object_not_exists",
				&qserror.QingStorError{
					StatusCode:   404,
					Code:         "object_not_exists",
					Message:      "",
					RequestID:    "",
					ReferenceURL: "",
				},
				services.ErrObjectNotExist,
			},
		}

		for _, tt := range tests {
			t.Run(tt.name, func(t *testing.T) {
				assert.True(t, errors.Is(formatError(tt.input), tt.expected))
			})
		}
	}
}

func Test_isObjectDirectory(t *testing.T) {
	tests := []struct {
		name    string
		keyType string
		want    bool
	}{
		{
			name:    "false",
			keyType: "application/octet-stream",
			want:    false,
		},
		{
			name:    "true",
			keyType: "application/x-directory",
			want:    true,
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			o := &service.KeyType{
				MimeType: convert.String(tt.keyType),
			}
			if got := isObjectDirectory(o); got != tt.want {
				t.Errorf("isObjectDirectory() = %v, want %v", got, tt.want)
			}
		})
	}
}


errors.go
package qingstor

import "github.com/beyondstorage/go-storage/v4/services"

var (
	// ErrBucketNameInvalid will be returned while bucket name is invalid.
	ErrBucketNameInvalid = services.NewErrorCode("invalid bucket name")

	// ErrWorkDirInvalid will be returned while work dir is invalid.
	// Work dir must start and end with only one '/'
	ErrWorkDirInvalid = services.NewErrorCode("invalid work dir")

	// ErrEncryptionCustomerKeyInvalid will be returned while encryption customer key is invalid.
	// Encryption key must be a 32-byte AES-256 key.
	ErrEncryptionCustomerKeyInvalid = services.NewErrorCode("invalid encryption customer key")

	// ErrAppendNextPositionEmpty will be returned while next append position is empty.
	ErrAppendNextPositionEmpty = services.NewErrorCode("next append position is empty")

	// ErrPartNumberInvalid will be returned while part number is out of range [0, 10000] when uploading multipart.
	ErrPartNumberInvalid = services.NewErrorCode("part number is out of range [0, 10000]")
)


service.go
package qingstor

import (
	"context"

	"github.com/qingstor/qingstor-sdk-go/v4/service"

	ps "github.com/beyondstorage/go-storage/v4/pairs"
	. "github.com/beyondstorage/go-storage/v4/types"
)

func (s *Service) create(ctx context.Context, name string, opt pairServiceCreate) (store Storager, err error) {
	// ServicePairCreate requires location, so we don't need to add location into pairs
	pairs := append(opt.pairs, ps.WithName(name))

	st, err := s.newStorage(pairs...)
	if err != nil {
		return
	}

	_, err = st.bucket.PutWithContext(ctx)
	if err != nil {
		return
	}
	return st, nil
}

func (s *Service) delete(ctx context.Context, name string, opt pairServiceDelete) (err error) {
	pairs := append(opt.pairs, ps.WithName(name))

	store, err := s.newStorage(pairs...)
	if err != nil {
		return
	}
	_, err = store.bucket.DeleteWithContext(ctx)
	if err != nil {
		return
	}
	return nil
}

func (s *Service) get(ctx context.Context, name string, opt pairServiceGet) (store Storager, err error) {
	pairs := append(opt.pairs, ps.WithName(name))

	store, err = s.newStorage(pairs...)
	if err != nil {
		return
	}
	return
}

func (s *Service) list(ctx context.Context, opt pairServiceList) (it *StoragerIterator, err error) {
	input := &storagePageStatus{}

	if opt.HasLocation {
		input.location = opt.Location
	}

	return NewStoragerIterator(ctx, s.nextStoragePage, input), nil
}

func (s *Service) nextStoragePage(ctx context.Context, page *StoragerPage) error {
	input := page.Status.(*storagePageStatus)

	serviceInput := &service.ListBucketsInput{
		Limit:  &input.offset,
		Offset: &input.limit,
	}
	if input.location != "" {
		serviceInput.Location = &input.location
	}

	output, err := s.service.ListBucketsWithContext(ctx, serviceInput)
	if err != nil {
		return err
	}

	for _, v := range output.Buckets {
		store, err := s.newStorage(ps.WithName(*v.Name), ps.WithLocation(*v.Location))
		if err != nil {
			return err
		}
		page.Data = append(page.Data, store)
	}

	input.offset += len(output.Buckets)
	if input.offset >= service.IntValue(output.Count) {
		return IterateDone
	}

	return nil
}


iterator.go
package qingstor

import (
	"strconv"
)

type objectPageStatus struct {
	delimiter string
	limit     int
	marker    string
	prefix    string
}

func (i *objectPageStatus) ContinuationToken() string {
	return i.marker
}

type segmentPageStatus struct {
	delimiter      string
	keyMarker      string
	limit          int
	prefix         string
	uploadIdMarker string
}

func (i *segmentPageStatus) ContinuationToken() string {
	return i.keyMarker + "/" + i.uploadIdMarker
}

type storagePageStatus struct {
	limit    int
	offset   int
	location string
}

func (i *storagePageStatus) ContinuationToken() string {
	return strconv.FormatInt(int64(i.offset), 10)
}

type partPageStatus struct {
	prefix           string
	limit            int
	partNumberMarker int
	uploadID         string
}

func (i *partPageStatus) ContinuationToken() string {
	return strconv.FormatInt(int64(i.partNumberMarker), 10)
}


utils.go
package qingstor

import (
	"crypto/md5"
	"encoding/base64"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"regexp"
	"strings"
	"time"

	"github.com/pengsrc/go-shared/convert"
	qsconfig "github.com/qingstor/qingstor-sdk-go/v4/config"
	iface "github.com/qingstor/qingstor-sdk-go/v4/interface"
	qserror "github.com/qingstor/qingstor-sdk-go/v4/request/errors"
	"github.com/qingstor/qingstor-sdk-go/v4/service"

	"github.com/beyondstorage/go-endpoint"
	ps "github.com/beyondstorage/go-storage/v4/pairs"
	"github.com/beyondstorage/go-storage/v4/pkg/credential"
	"github.com/beyondstorage/go-storage/v4/pkg/headers"
	"github.com/beyondstorage/go-storage/v4/pkg/httpclient"
	"github.com/beyondstorage/go-storage/v4/services"
	typ "github.com/beyondstorage/go-storage/v4/types"
)

// Service is the qingstor service config.
type Service struct {
	config  *qsconfig.Config
	service iface.Service

	client *http.Client

	defaultPairs DefaultServicePairs
	features     ServiceFeatures

	typ.UnimplementedServicer
}

// String implements Service.String.
func (s *Service) String() string {
	if s.config == nil {
		return fmt.Sprintf("Servicer qingstor")
	}
	return fmt.Sprintf("Servicer qingstor {AccessKey: %s}", s.config.AccessKeyID)
}

// Storage is the qingstor object storage client.
type Storage struct {
	bucket     iface.Bucket
	config     *qsconfig.Config
	properties *service.Properties

	defaultPairs DefaultStoragePairs
	features     StorageFeatures

	// options for this storager.
	workDir string // workDir dir for all operation.

	typ.UnimplementedStorager
	typ.UnimplementedCopier
	typ.UnimplementedFetcher
	typ.UnimplementedMover
	typ.UnimplementedMultiparter
	typ.UnimplementedReacher
	typ.UnimplementedAppender
	typ.UnimplementedDirer
}

// String implements Storager.String
func (s *Storage) String() string {
	// qingstor work dir should start and end with "/"
	return fmt.Sprintf(
		"Storager qingstor {Name: %s, Location: %s, WorkDir: %s}",
		*s.properties.BucketName, *s.properties.Zone, s.workDir,
	)
}

// New will create both Servicer and Storager.
func New(pairs ...typ.Pair) (typ.Servicer, typ.Storager, error) {
	return newServicerAndStorager(pairs...)
}

// NewServicer will create Servicer only.
func NewServicer(pairs ...typ.Pair) (typ.Servicer, error) {
	return newServicer(pairs...)
}

// NewStorager will create Storager only.
func NewStorager(pairs ...typ.Pair) (typ.Storager, error) {
	_, store, err := newServicerAndStorager(pairs...)
	return store, err
}

func newServicer(pairs ...typ.Pair) (srv *Service, err error) {
	defer func() {
		if err != nil {
			err = services.InitError{Op: "new_servicer", Type: Type, Err: formatError(err), Pairs: pairs}
		}
	}()

	opt, err := parsePairServiceNew(pairs)
	if err != nil {
		return nil, err
	}

	srv = &Service{
		client: httpclient.New(opt.HTTPClientOptions),
	}

	var cfg *qsconfig.Config

	// Set config's credential.
	cp, err := credential.Parse(opt.Credential)
	if err != nil {
		return nil, err
	}
	switch cp.Protocol() {
	case credential.ProtocolHmac:
		cfg, err = qsconfig.New(cp.Hmac())
		if err != nil {
			return nil, err
		}
	default:
		return nil, services.PairUnsupportedError{Pair: ps.WithCredential(opt.Credential)}
	}

	// Set config's endpoint
	if opt.HasEndpoint {
		ep, err := endpoint.Parse(opt.Endpoint)
		if err != nil {
			return nil, err
		}

		switch ep.Protocol() {
		case endpoint.ProtocolHTTPS:
			_, cfg.Host, cfg.Port = ep.HTTPS()
		case endpoint.ProtocolHTTP:
			_, cfg.Host, cfg.Port = ep.HTTP()
		default:
			return nil, services.PairUnsupportedError{Pair: ps.WithEndpoint(opt.Endpoint)}
		}

		cfg.Protocol = ep.Protocol()
	}
	// Set config's http client
	cfg.Connection = srv.client

	srv.config = cfg
	srv.service, _ = service.Init(cfg)

	if opt.HasDefaultServicePairs {
		srv.defaultPairs = opt.DefaultServicePairs
	}
	if opt.HasServiceFeatures {
		srv.features = opt.ServiceFeatures
	}
	return
}

// New will create a new qingstor service.
func newServicerAndStorager(pairs ...typ.Pair) (srv *Service, store *Storage, err error) {
	srv, err = newServicer(pairs...)
	if err != nil {
		return
	}

	store, err = srv.newStorage(pairs...)
	if err != nil {
		err = services.InitError{Op: "new_storager", Type: Type, Err: formatError(err), Pairs: pairs}
		return
	}
	return
}

// multipartXXX are multipart upload restriction in QingStor, see more detail at:
// https://docs.qingcloud.com/qingstor/api/object/multipart/index.html#%E5%88%86%E6%AE%B5%E4%B8%8A%E4%BC%A0%E9%99%90%E5%88%B6
const (
	// multipartNumberMinimum is the min part count supported
	multipartNumberMinimum = 0
	// multipartNumberMaximum is the max part count supported
	multipartNumberMaximum = 10000
	// multipartSizeMaximum is the maximum size for each part, 5GB
	multipartSizeMaximum = 5 * 1024 * 1024 * 1024
	// multipartSizeMinimum is the minimum size for each part, except the last part, 4MB
	multipartSizeMinimum = 4 * 1024 * 1024
)

const (
	// writeSizeMaximum is the maximum size for write operation, 5GB.
	// ref: https://docs.qingcloud.com/qingstor/#object
	writeSizeMaximum = 5 * 1024 * 1024 * 1024
	// copySizeMaximum is the maximum size for copy operation, 5GB.
	// ref: https://docs.qingcloud.com/qingstor/api/object/copy
	copySizeMaximum = 5 * 1024 * 1024 * 1024
	// appendSizeMaximum is the maximum append size for per append operation, 5GB.
	// ref: https://docs.qingcloud.com/qingstor/api/object/append
	appendSizeMaximum = 5 * 1024 * 1024 * 1024
	// appendSizeMaximum is the total maximum size for an append object, 5TB.
	// ref: https://docs.qingcloud.com/qingstor/api/object/append
	appendTotalSizeMaximum = 50 * 1024 * 1024 * 1024 * 1024
)

// bucketNameRegexp is the bucket name regexp, which indicates:
// 1. length: 6-63;
// 2. contains lowercase letters, digits and strikethrough;
// 3. starts and ends with letter or digit.
var bucketNameRegexp = regexp.MustCompile(`^[a-z\d][a-z-\d]{4,61}[a-z\d]$`)

// IsBucketNameValid will check whether given string is a valid bucket name.
func IsBucketNameValid(s string) bool {
	return bucketNameRegexp.MatchString(s)
}

func formatError(err error) error {
	if _, ok := err.(services.InternalError); ok {
		return err
	}

	// Handle errors returned by qingstor.
	var e *qserror.QingStorError
	if !errors.As(err, &e) {
		return fmt.Errorf("%w: %v", services.ErrUnexpected, err)
	}

	switch e.Code {
	case "":
		// code=="" means this response doesn't have body.
		switch e.StatusCode {
		case 404:
			return fmt.Errorf("%w: %v", services.ErrObjectNotExist, e)
		default:
			return e
		}
	case "permission_denied":
		return fmt.Errorf("%w: %v", services.ErrPermissionDenied, e)
	case "object_not_exists":
		return fmt.Errorf("%w: %v", services.ErrObjectNotExist, e)
	default:
		return fmt.Errorf("%w: %v", services.ErrUnexpected, err)
	}
}

func convertUnixTimestampToTime(v int) time.Time {
	if v == 0 {
		return time.Time{}
	}
	return time.Unix(int64(v), 0)
}

// All available storage classes are listed here.
const (
	StorageClassStandard   = "STANDARD"
	StorageClassStandardIA = "STANDARD_IA"
)

func (s *Service) newStorage(pairs ...typ.Pair) (store *Storage, err error) {
	opt, err := parsePairStorageNew(pairs)
	if err != nil {
		return
	}

	// WorkDir should be an abs path, start and ends with "/"
	if opt.HasWorkDir && !isWorkDirValid(opt.WorkDir) {
		err = ErrWorkDirInvalid
		return
	}
	// set work dir into root path if no work dir passed
	if !opt.HasWorkDir {
		opt.WorkDir = "/"
	}

	if !IsBucketNameValid(opt.Name) {
		err = ErrBucketNameInvalid
		return
	}

	// Detect location automatically
	if !opt.HasLocation {
		opt.Location, err = s.detectLocation(opt.Name)
		if err != nil {
			return
		}
	}

	bucket, err := s.service.Bucket(opt.Name, opt.Location)
	if err != nil {
		return
	}

	st := &Storage{
		bucket:     bucket,
		config:     bucket.Config,
		properties: bucket.Properties,

		workDir: "/",
	}

	if opt.HasDisableURICleaning {
		st.config.DisableURICleaning = opt.DisableURICleaning
	}
	if opt.HasDefaultStoragePairs {
		st.defaultPairs = opt.DefaultStoragePairs
	}
	if opt.HasStorageFeatures {
		st.features = opt.StorageFeatures
	}
	if opt.HasWorkDir {
		st.workDir = opt.WorkDir
	}
	return st, nil
}

func (s *Service) detectLocation(name string) (location string, err error) {
	defer func() {
		err = s.formatError("detect_location", err, "")
	}()

	u := fmt.Sprintf("%s://%s:%d/%s", s.config.Protocol, s.config.Host, s.config.Port, name)

	r, err := s.client.Head(u)
	if err != nil {
		return
	}
	if r.StatusCode != http.StatusMovedPermanently {
		err = fmt.Errorf("%w: head status is %d instead of %d", services.ErrUnexpected, r.StatusCode, http.StatusMovedPermanently)
		return
	}

	// Example URL: https://zone.qingstor.com/bucket
	locationUrl, err := url.Parse(r.Header.Get(headers.Location))
	if err != nil {
		return
	}
	location = strings.Split(locationUrl.Host, ".")[0]
	return
}

func (s *Service) formatError(op string, err error, name string) error {
	if err == nil {
		return nil
	}

	return services.ServiceError{
		Op:       op,
		Err:      formatError(err),
		Servicer: s,
		Name:     name,
	}
}

// isWorkDirValid check qingstor work dir
// work dir must start with only one "/" (abs path), and end with only one "/" (a dir).
// If work dir is the root path, set it to "/".
func isWorkDirValid(wd string) bool {
	return strings.HasPrefix(wd, "/") && // must start with "/"
		strings.HasSuffix(wd, "/") && // must end with "/"
		!strings.HasPrefix(wd, "//") && // not start with more than one "/"
		!strings.HasSuffix(wd, "//") // not end with more than one "/"
}

// getAbsPath will calculate object storage's abs path
func (s *Storage) getAbsPath(path string) string {
	prefix := strings.TrimPrefix(s.workDir, "/")
	return prefix + path
}

// getRelPath will get object storage's rel path.
func (s *Storage) getRelPath(path string) string {
	prefix := strings.TrimPrefix(s.workDir, "/")
	return strings.TrimPrefix(path, prefix)
}

func (s *Storage) formatError(op string, err error, path ...string) error {
	if err == nil {
		return nil
	}

	return services.StorageError{
		Op:       op,
		Err:      formatError(err),
		Storager: s,
		Path:     path,
	}
}

func (s *Storage) newObject(done bool) *typ.Object {
	return typ.NewObject(s, done)
}

func (s *Storage) formatFileObject(v *service.KeyType) (o *typ.Object, err error) {
	o = s.newObject(false)
	o.ID = *v.Key
	o.Path = s.getRelPath(*v.Key)
	o.Mode |= typ.ModeRead

	o.SetContentLength(service.Int64Value(v.Size))
	o.SetLastModified(convertUnixTimestampToTime(service.IntValue(v.Modified)))

	if v.MimeType != nil {
		o.SetContentType(service.StringValue(v.MimeType))
	}
	if v.Etag != nil {
		o.SetEtag(service.StringValue(v.Etag))
	}

	var sm ObjectSystemMetadata
	if value := service.StringValue(v.StorageClass); value != "" {
		sm.StorageClass = value
	}
	o.SetSystemMetadata(sm)

	return o, nil
}

func isObjectDirectory(o *service.KeyType) bool {
	return convert.StringValue(o.MimeType) == "application/x-directory"
}

// All available SSE customer algorithms are listed here.
const (
	SseCustomerAlgorithmAes256 = "AES256"
)

func calculateEncryptionHeaders(algo string, key []byte) (algorithm, keyBase64, keyMD5Base64 *string, err error) {
	if len(key) != 32 {
		err = ErrEncryptionCustomerKeyInvalid
		return
	}
	kB64 := base64.StdEncoding.EncodeToString(key)
	kMD5 := md5.Sum(key)
	kMD5B64 := base64.StdEncoding.EncodeToString(kMD5[:])
	return &algo, &kB64, &kMD5B64, nil
}
